internal/mturk/step.go

```diff
@@ -189,19 +191,135 @@ func (s *Session) runMTurkMessageStep(ctx context.Context, project *ent.Project,
+		WithRun(func(run *ent.RunQuery) {
+			run.WithSteps()
+		}).
+		Where(stepRunModel.IDEQ(stepRun.ID)).
+		First(ctx)
+	if err != nil {
+		log.Error().Err(err).Msg("get stepRun")
+		return errors.Wrap(err, "get stepRun")
+	}
+
+	step, err = stepRun.Edges.StepOrErr()
+	if err != nil {
+		log.Error().Err(err).Msg("get step")
+		return errors.Wrap(err, "get step")
+	}
+
+	run, err = stepRun.Edges.RunOrErr()
+	if err != nil {
+		log.Error().Err(err).Msg("get run")
+		return errors.Wrap(err, "get run")
+	}
+
+	stepRuns, err := run.Edges.StepsOrErr()
+	if err != nil {
+		log.Error().Err(err).Msg("get step runs")
+		return errors.Wrap(err, "get step runs")
+	}
+
+	template, err = step.Edges.TemplateOrErr()
+	if err != nil {
+		log.Error().Err(err).Msg("get template")
+		return errors.Wrap(err, "get template")
 	}
 
-	output, err := s.notifyWorkers(ctx, subject, msg, workerIDs)
+	steps, err := template.Edges.StepsOrErr()
 	if err != nil {
-		return errors.Wrap(err, "notify workers")
+		log.Error().Err(err).Msg("get steps")
+		return errors.Wrap(err, "get steps")
 	}
 
+	rsteps := make([]*model.RenderStep, len(steps))
+	t := *run.StartedAt
+	for i, s := range steps {
+		var startsAt, startedAt, endedAt string
+		if stepRuns[i].Index <= stepRun.Index {
+			startedAt = stepRun.StartedAt.Format(time.Kitchen)
+			startsAt = startedAt
+			if stepRuns[i].Index < stepRun.Index {
+				endedAt = stepRun.EndedAt.Format(time.Kitchen)
+			}
+		} else {
+			startsAt = t.Format(time.Kitchen)
+			t = t.Add(time.Duration(step.Duration) * time.Minute)
+		}
+		rsteps[i] = &model.RenderStep{
+			Index:             s.Index,
+			Duration:          s.Duration,
+			ParticipantsCount: stepRuns[i].ParticipantsCount,
+			Type:              s.Type.String(),
+			StartsAt:          startsAt,
+			StartedAt:         startedAt,
+			EndedAt:           endedAt,
+		}
+	}
+
+	startedAt := stepRun.StartedAt.Format(time.Kitchen)
 	failedWorkedIDs := make(map[string]*mturk.NotifyWorkersFailureStatus)
-	for _, stat := range output.NotifyWorkersFailureStatuses {
-		failedWorkedIDs[*stat.WorkerId] = stat
+	for _, workerID := range workerIDs {
+		tempWorkerIDs := make([]string, 0, 1)
+		tempWorkerIDs = append(tempWorkerIDs, workerID)
+
+		renderCtx := &model.RenderContext{
+			Template: &model.RenderTemplate{
+				Adult:            template.Adult,
+				Sandbox:          template.Sandbox,
+				SelectionType:    template.SelectionType.String(),
+				Name:             template.Name,
+				ParticipantCount: template.ParticipantCount,
+			},
+			Run: &model.RenderRun{
+				Name:      run.Name,
+				StartedAt: run.StartedAt.Format(time.Kitchen),
+			},
+			Step: &model.RenderStep{
+				Index:             step.Index,
+				Duration:          step.Duration,
+				ParticipantsCount: stepRun.ParticipantsCount,
+				Type:              step.Type.String(),
+				StartsAt:          startedAt,
+				StartedAt:         startedAt,
+			},
+			Steps:    rsteps,
+			URL:      url.String(),
+			WorkerID: workerID,
+		}
```

Comment:
The creation of the renderCtx here is still happening over and over, while the only changed fields from participant to participant are url and workerID. Create renderCtx outside the workers loop, and on change those 2 fields inside the loop. Actually, only only WorkerID for now, we'll later customize the URL per participant and we can make that field different per participant then.